lib/Conversion/ImportVerilog/FormatStrings.cpp

```diff
@@ -0,0 +1,200 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "ImportVerilogInternals.h"
+#include "slang/text/SFormat.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace ImportVerilog;
+using moore::IntAlign;
+using moore::IntFormat;
+using moore::IntPadding;
+using slang::SFormat::FormatOptions;
+
+namespace {
+struct FormatStringParser {
+  Context &context;
+  OpBuilder &builder;
+  /// The remaining arguments to be parsed.
+  ArrayRef<const slang::ast::Expression *> arguments;
+  /// The current location to use for ops and diagnostics.
+  Location loc;
+  /// The default format for integer arguments not covered by a format string
+  /// literal.
+  IntFormat defaultFormat;
+  /// The interpolated string fragments that will be concatenated using a
+  /// `moore.fmt.concat` op.
+  SmallVector<Value> fragments;
+
+  FormatStringParser(Context &context,
+                     ArrayRef<const slang::ast::Expression *> arguments,
+                     Location loc, IntFormat defaultFormat)
+      : context(context), builder(context.builder), arguments(arguments),
+        loc(loc), defaultFormat(defaultFormat) {}
+
+  /// Entry point to the format string parser.
+  FailureOr<Value> parse(bool appendNewline) {
+    while (!arguments.empty()) {
+      const auto &arg = *arguments[0];
+      arguments = arguments.drop_front();
+      if (arg.kind == slang::ast::ExpressionKind::EmptyArgument)
+        continue;
+      loc = context.convertLocation(arg.sourceRange);
+      if (auto *lit = arg.as_if<slang::ast::StringLiteral>()) {
+        if (failed(parseFormat(lit->getValue())))
+          return failure();
+      } else {
+        if (failed(emitDefault(arg)))
+          return failure();
+      }
+    }
+
+    // Append the optional newline.
+    if (appendNewline)
+      emitLiteral("\n");
+
+    // Concatenate all string fragments into one formatted string, or return an
+    // empty literal if no fragments were generated.
+    if (fragments.empty())
+      return Value{};
+    if (fragments.size() == 1)
+      return fragments[0];
+    return builder.create<moore::FormatConcatOp>(loc, fragments).getResult();
+  }
+
+  /// Parse a format string literal and consume and format the arguments
+  /// corresponding to the format specifiers it contains.
+  LogicalResult parseFormat(StringRef format) {
+    bool anyFailure = false;
+    auto onText = [&](auto text) {
+      if (anyFailure)
+        return;
+      emitLiteral(text);
+    };
+    auto onArg = [&](auto specifier, auto offset, auto len,
+                     const auto &options) {
+      if (anyFailure)
+        return;
+      if (failed(emitArgument(specifier, format.substr(offset, len), options)))
+        anyFailure = true;
+    };
+    auto onError = [&](auto, auto, auto, auto) {};
```

Comment:
Does `onError` need to set `anyFailure` when invoked?

Reply:
The Slang frontend already runs this `parse` function and reports any errors it encounters. When ImportVerilog gets to this point, all errors have already reported and compilation would have already been aborted. I'll add a comment, and an assertion inside `onError` to check that this is correct. Thanks for the pointer!